compiler/src/iree/compiler/Preprocessing/Common/ApplyPDLPatterns.cpp

```diff
@@ -0,0 +1,529 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Preprocessing/Common/Passes.h"
+
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "iree/compiler/Dialect/Stream/IR/StreamOps.h"
+#include "iree/compiler/Dialect/Util/IR/UtilDialect.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/SourceMgr.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/PDL/IR/PDL.h"
+#include "mlir/Dialect/PDLInterp/IR/PDLInterp.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Utils/StaticValueUtils.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Parser/Parser.h"
+#include "mlir/Support/FileUtilities.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-preprocessing-apply-pdl-patterns"
+
+using namespace mlir;
+using namespace mlir::iree_compiler;
+
+namespace mlir::iree_compiler::Preprocessing {
+
+#define GEN_PASS_DEF_APPLYPDLPATTERNS
+#include "iree/compiler/Preprocessing/Common/Passes.h.inc" // IWYU pragma: export
+
+} // namespace mlir::iree_compiler::Preprocessing
+
+// Native rewrite function used for the rewrite
+LogicalResult rewriteAsDispatch(PatternRewriter &rewriter, PDLResultList &,
+                                ArrayRef<PDLValue> values) {
+  Operation *op = values[0].cast<Operation *>();
+  rewriter.replaceOp(op, op->getOperand(0));
+  return success();
+}
+
+// Get the `memref` type for a `tensor` type.
+static MemRefType getMemRefTypeFor(RankedTensorType tensorType) {
+  return MemRefType::get(tensorType.getShape(), tensorType.getElementType());
+}
+
+// Generate the external function call type that corresponds to the
+// matched list.
+static FunctionType getExternalFunctionCallType(MLIRContext *context,
+                                                Location loc,
+                                                TypeRange inputTypes,
+                                                TypeRange resultTypes,
+                                                TypeRange otherOperandTypes) {
+  SmallVector<Type> externalCallArgTypes;
+  // Conversion from tensor types to call arg types.
+  auto convertTensorTypeToCallArgTypes = [&](RankedTensorType tensorType) {
+    auto memRefType = getMemRefTypeFor(tensorType);
+    externalCallArgTypes.push_back(
+        MemRefType::get(ArrayRef<int64_t>{}, memRefType.getElementType()));
+    externalCallArgTypes.push_back(IndexType::get(context));
+  };
+
+  // Conversion from input type to call arg types.
+  auto convertInputTypeToCallArgTypes = [&](Type inputType) {
+    if (inputType.isIntOrFloat()) {
+      externalCallArgTypes.push_back(inputType);
+      return;
+    }
+
+    if (auto tensorType = inputType.dyn_cast<RankedTensorType>()) {
+      convertTensorTypeToCallArgTypes(tensorType);
+      return;
+    }
+
+    llvm_unreachable(
```

Comment:
properly error out instead (we don't want users passing in pdl specs to the release compiler to get undefined behavior/crashes/etc)

(here and all other llvm_unreachable that come from user inputs)